executor/new_index_lookup_join.go

```diff
@@ -0,0 +1,534 @@
+	innerCh  chan<- *lookUpJoinTask
+}
+
+type innerWorker struct {
+	innerCtx
+
+	taskCh      <-chan *lookUpJoinTask
+	outerCtx    outerCtx
+	ctx         context.Context
+	executorChk *chunk.Chunk
+}
+
+// Open implements the Executor interface.
+func (e *NewIndexLookUpJoin) Open(goCtx goctx.Context) error {
+	err := e.children[0].Open(goCtx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.startWorkers(goCtx)
+	return nil
+}
+
+func (e *NewIndexLookUpJoin) startWorkers(goCtx goctx.Context) {
+	concurrency := e.ctx.GetSessionVars().IndexLookupConcurrency
+	resultCh := make(chan *lookUpJoinTask, concurrency)
+	e.resultCh = resultCh
+	workerCtx, cancelFunc := goctx.WithCancel(goCtx)
+	e.cancelFunc = cancelFunc
+	innerCh := make(chan *lookUpJoinTask, concurrency)
+	e.workerWg.Add(1)
+	go e.newOuterWorker(resultCh, innerCh).run(workerCtx, e.workerWg)
+	e.workerWg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		go e.newInnerWorker(innerCh).run(workerCtx, e.workerWg)
+	}
+}
+
+func (e *NewIndexLookUpJoin) newOuterWorker(resultCh, innerCh chan *lookUpJoinTask) *outerWorker {
+	ow := &outerWorker{
+		outerCtx:     e.outerCtx,
+		ctx:          e.ctx,
+		executor:     e.children[0],
+		executorChk:  chunk.NewChunk(e.outerCtx.rowTypes),
+		resultCh:     resultCh,
+		innerCh:      innerCh,
+		batchSize:    32,
+		maxBatchSize: e.ctx.GetSessionVars().IndexLookupSize,
+	}
+	return ow
+}
+
+func (e *NewIndexLookUpJoin) newInnerWorker(taskCh chan *lookUpJoinTask) *innerWorker {
+	iw := &innerWorker{
+		innerCtx:    e.innerCtx,
+		outerCtx:    e.outerCtx,
+		taskCh:      taskCh,
+		ctx:         e.ctx,
+		executorChk: chunk.NewChunk(e.innerCtx.rowTypes),
+	}
+	return iw
+}
+
+// NextChunk implements the Executor interface.
+func (e *NewIndexLookUpJoin) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	for {
+		err := e.prepareJoinResult(goCtx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if e.joinResult.NumRows() == 0 {
+			return nil
+		}
+		for e.joinResultCursor < e.joinResult.NumRows() {
+			if chk.NumRows() == e.maxChunkSize {
+				return nil
+			}
+			chk.AppendRow(0, e.joinResult.GetRow(e.joinResultCursor))
+			e.joinResultCursor++
+		}
+	}
+}
+
+// Next implements the Executor interface.
+// Even though we only support read children in chunk mode, but we are not sure if our parents
+// support chunk, so we have to implement Next.
+func (e *NewIndexLookUpJoin) Next(goCtx goctx.Context) (Row, error) {
+	for {
+		err := e.prepareJoinResult(goCtx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if e.joinResult.NumRows() == 0 {
+			return nil, nil
+		}
+		row := e.joinResult.GetRow(e.joinResultCursor)
+		datumRow := make(types.DatumRow, row.Len())
+		for i := 0; i < row.Len(); i++ {
+			datumRow[i] = row.GetDatum(i, e.schema.Columns[i].RetType)
+		}
+		e.joinResultCursor++
+		return datumRow, nil
+	}
+}
+
+func (e *NewIndexLookUpJoin) prepareJoinResult(goCtx goctx.Context) error {
```

Comment:
s/prepareJoinResult/getJoinResult/

Reply:
But this function doesn't return the joinResult.